lib/Conversion/BGVToPoly/BGVToPoly.cpp

```diff
@@ -0,0 +1,103 @@
+#include "include/Conversion/BGVToPoly/BGVToPoly.h"
+
+#include <cstddef>
+
+#include "include/Dialect/BGV/IR/BGVDialect.h"
+#include "include/Dialect/BGV/IR/BGVOps.h"
+#include "include/Dialect/BGV/IR/BGVTypes.h"
+#include "include/Dialect/Poly/IR/PolyAttributes.h"
+#include "include/Dialect/Poly/IR/PolyOps.h"
+#include "include/Dialect/Poly/IR/PolyTypes.h"
+#include "include/Dialect/Poly/IR/Polynomial.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/OneToNFuncConversions.h"
+#include "mlir/Transforms/OneToNTypeConversion.h"
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/OneToNTypeConversion.h"
+
+namespace mlir::heir::bgv {
+
+#define GEN_PASS_DEF_BGVTOPOLY
+#include "include/Conversion/BGVToPoly/BGVToPoly.h.inc"
+
+class CiphertextTypeConverter : public OneToNTypeConverter {
+ public:
+  CiphertextTypeConverter() {
+    addConversion([](Type type) { return type; });
+    addConversion([](CiphertextType type, SmallVectorImpl<Type> &results) {
+      auto ring = type.getRings().getRings()[0]; // using the first ring.
+      // TODO: figure out the right way to select ring;
+      auto polyType = poly::PolynomialType::get(type.getContext(), ring);
+
+      for (size_t i = 0; i < type.getDim(); i++) {
+        results.push_back(polyType);
+      }
+      return success();
+    });
+  }
+};
+
+struct ConvertAdd : public OneToNOpConversionPattern<AddOp> {
+  using OneToNOpConversionPattern<AddOp>::OneToNOpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      AddOp op, OpAdaptor adaptor,
+      OneToNPatternRewriter &rewriter) const override {
+    auto xy = adaptor.getOperands();
+    if (xy.size() != 2) {
+      op->emitError() << "Unexpected number of operands, expected 2, got "
+                      << xy.size();
+      return failure();
+    }
+
+    auto x = xy[0];
+    auto y = xy[1];
+    if (x.size() != y.size()) {
+      op.emitError() << "Operands have different number of polys " << x.size()
+                     << " != " << y.size();
+      return failure();
+    }
+
+    SmallVector<Value> z;
+    for (size_t i = 0; i < x.size(); ++i) {
+      auto xt = x[i].getType();
+      auto yt = y[i].getType();
+      if (xt != yt) {
+        op->emitError() << "Mismatch of polynomial types, got " << xt << " and " << yt;
+        return failure();
+      }
+      z.push_back(rewriter.create<poly::AddOp>(op.getLoc(), xt,
```

Comment:
I see your point (2), though I assume that set of OPs independent argument-wise from each other should "look parallelize-able" to a downstream passes as well as a loop. Do we have any info about those downstream passes, what can and can not be parallelized?

For (1), there is a difficulty with IR, polys are separate values, IIUC to use them in a loop  one need to organize them in the collection of some sort that supports indexing, but MLIR builtin doesn't provide such collection type. Do you have any suggestions on how to make it work?

Reply:
> MLIR builtin doesn't provide such collection type

What do you think about lowering a BGV ciphertext to a tensor of polys? I think this would be a more natural collection for them.

> I assume that set of OPs independent argument-wise from each other should "look parallelize-able" to a downstream passes as well as a loop.

Yeah, I think you are right here - the IR is in graph representation, and so it should be easy for lower levels to sort and deduce parallel tasks - so I think this is not as much of a reason to do it. Using a tensor, however, may be.

Reply:
What would be a benefit of using `tensor`  instead of flat `poly`s?
Some disadvantage is 4 times increase in IR size:
```mlir
....
%i2 = arith.constant 2 : i64
%l2 = tensor.extract %l[%ii] : !poly.poly<...>
%r2 = tensor.extract %r[%ii] : !poly.poly<...>
%s2 = poly.add(%l2, %r2) : !poly.poly<...>
....
%s = tensor.from_elements ..., %s2, ... : tensor<3x!poly.poly<...>>
return %s: tensor<3x!poly.poly<...>>
```
instead of 
```mlir
%s2 = poly.add(%l2, %r2) : !poly.poly<...>
...
return ..., %s2, ... : ..., !poly.poly<...>, ...
```
Of course some ops (such as `Add`, `Sub`, `Negate`, and `ModuleSwitch`) could be made shorter by using loops, while others don't.
Also number of polys (is my understanding about number of polys correct) is ~ explosion rate of tensor/flat IR (=4) => no clear "readability gains".

Reply:
@j2kun helped point me to some resources about this, add anything else that there is!

We were looking aorund at what MLIR provides - it turns out if you add `[ElementwiseMappable](https://mlir.llvm.org/docs/Traits/#elementwisemappable)` to the `poly` ops that work element-wise, then you can also apply these ops onto tensors of the types as well (which means for the simple ops, you won't need to deal with extracting elements OR dealing with the for loops). 

But without considering IR size, maintaining ciphertext structure might give us some advantages without losing out (first, we don't lose out that much because if we maintain structure, we can always detensorize ops later). But if ops use tensors and use the ElementwiseMapping trait above, we can also use existing passes https://mlir.llvm.org/docs/Passes/#-convert-elementwise-to-linalg to create parallel for-loops (which at this level would save on a factor of two / dimension-size when lowering to further parallel ops like adds).